cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java

```diff
@@ -127,24 +133,44 @@ public void startTx(Transaction tx) {
 
   @Override
   public List<Map<byte[], byte[]>> getPersisted(List<co.cask.cdap.api.dataset.table.Get> gets) {
-    List<Get> hbaseGets = Lists.transform(gets, new Function<co.cask.cdap.api.dataset.table.Get, Get>() {
-      @Nullable
-      @Override
-      public Get apply(co.cask.cdap.api.dataset.table.Get get) {
-        List<byte[]> cols = get.getColumns();
-        return createGet(get.getRow(), cols == null ? null : cols.toArray(new byte[cols.size()][]));
+    List<Get> hbaseGets = new ArrayList<>();
+    for (co.cask.cdap.api.dataset.table.Get get : gets) {
+      List<byte[]> cols = get.getColumns();
+      // Our get with empty array means get nothing. Theirs means get all.
+      // That's why we need this somewhat-messy code.
+      if (cols == null || !cols.isEmpty()) {
+        hbaseGets.add(createGet(get.getRow(), cols == null ? null : cols.toArray(new byte[cols.size()][])));
       }
-    });
+    }
+
     try {
-      Result[] results = hTable.get(hbaseGets);
-      return Lists.transform(Arrays.asList(results), new Function<Result, Map<byte[], byte[]>>() {
-        @Nullable
-        @Override
-        public Map<byte[], byte[]> apply(Result result) {
-          Map<byte[], byte[]> familyMap = result.getFamilyMap(columnFamily);
-          return familyMap != null ? familyMap : ImmutableMap.<byte[], byte[]>of();
+      List<Map<byte[], byte[]>> results =
```

Comment:
Is this needed? Can't you just apply the transform in the else branch of the next loop?